File: aoc-py/solutions/day4.py
"""
Day 4: Printing Department

https://adventofcode.com/2025/day/4
"""
__all__ = ('Day4',)

from typing import ClassVar

from ..solution import Solution
from ..utils import neighbors_4, neighbors_diag

class Day4(Solution):
    NAME: ClassVar[str] = 'Printing Department'

    def part_one(self, inp: str) -> int:
        grid = inp.splitlines()
        count = 0

        n_rows = len(grid)
        n_cols = len(grid[0])

        return sum(
            sum(
                i in range(n_rows) and j in range(n_cols) and grid[i][j] == '@'
                for i, j in neighbors_4(i, j) + neighbors_diag(i, j)
            ) < 4
            for i, row in enumerate(grid)
            for j, cell in enumerate(row)
            if cell == '@'
        )

    def part_two(self, inp: str) -> int:
        grid = [list(row) for row in inp.splitlines()]
        count = 0

        n_rows = len(grid)
        n_cols = len(grid[0])

        while True:
            accessible = 0

            for i, row in enumerate(grid):
                for j, cell in enumerate(row):
                    if cell == '@' and sum(
                        i in range(n_rows) and j in range(n_cols) and grid[i][j] == '@'
                        for i, j in neighbors_4(i, j) + neighbors_diag(i, j)
                    ) < 4:
                        grid[i][j] = '.'
                        accessible += 1

            if accessible == 0:
                break

            count += accessible
        return count

    def run(self, inp: str) -> None:
        print('Part 1:', p1 := self.part_one(inp))
        print('Part 2:', p2 := self.part_two(inp))

        assert p1 == 1460
        assert p2 == 9243

File: src/bin/day4.rs
//! Day 4: Printing Department
//!
//! <https://adventofcode.com/2025/day/4>
use std::fmt::Display;
use aoc_2025::{Solution, get_grid, neighbors};

pub struct Day4;

impl Solution for Day4 {
    const NAME: &'static str = "Printing Department";

    fn part_one<T: Display>(&self, inp: T) -> Self::OutputP1 {
        let grid = get_grid(inp);

        let n_rows = 0..grid.len();
        let n_cols  = 0..grid[0].len();

        grid
            .iter()
            .enumerate()
            .map(|(i, row)| row
                .iter()
                .enumerate()
                .filter(|&(j, &cell)| cell == b'@' && neighbors(i, j)
                    .iter()
                    .filter(|(i, j)| n_rows.contains(i) && n_cols.contains(j) && grid[*i][*j] == b'@')
                    .count() < 4
                )
                .count()
            )
            .sum()
    }

    fn part_two<T: Display>(&self, inp: T) -> Self::OutputP2 {
        let mut grid = get_grid(inp);

        let n_rows = grid.len();
        let n_cols  = grid[0].len();

        let mut count = 0;

        loop {
            let mut accessible = 0;

            for i in 0..n_rows {
                for j in 0..n_cols {
                    if grid[i][j] == b'@' && neighbors(i, j)
                        .iter()
                        .filter(|(i, j)|
                            (0..n_rows).contains(i)
                            && (0..n_cols).contains(j)
                            && grid[*i][*j] == b'@'
                        )
                        .count() < 4
                    {
                        grid[i][j] = b'.';
                        accessible += 1;
                    }
                }
            }

            if accessible == 0 {
                break;
            }
            count += accessible;
        }
        count
    }

    fn run(&self, inp: String) {
        let p1 = self.part_one(&inp);
        let p2 = self.part_two(&inp);

        println!("Part 1: {p1}");
        println!("Part 2: {p2}");

        assert_eq!(p1, 1460);
        assert_eq!(p2, 9243);
    }
}

fn main() {
    aoc_2025::run_day(4, &Day4);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() { main(); }
}